### PR TITLE
completion: Detect Homebrew prefix natively instead of idiomatically

### LINKED ISSUE #1978: Calling 'brew --prefix' is slow in the completions module - let's speed it up!
### Description

While profiling my prezto config, I noticed that the completions module was taking a bit longer than I felt it should (fractions of seconds, not whole seconds, but still it's slower than I wanted). After profiling, I determined that a subshell call to `$(brew --prefix)` was the culprit. I'm not the first to notice performance related to `brew --prefix`: https://github.com/Homebrew/brew/issues/3327

### Expected behavior

Since the completions module is one of the core Prezto modules and is enabled by default, I'd like to see it be as speedy as possible.

### Actual behavior

It's not speedy.

### Steps to Reproduce

1. On a Mac, install homebrew
2. Use brew to install curl (brew install curl)
3. Prezto now slows down because now the completions for curl are in a location managed by homebrew

### Versions

  - Prezto commit: latest (7884fb2271d131bb492fafc4eba1c537a821092b)
  - ZSH version: zsh 5.8 (x86_64-apple-darwin20.1.0) 
  - OS information: MacOS 12.2

### Potential solution

Current code:

```zsh
# Add completion for keg-only brewed curl when available.
if (( $+commands[brew] )) \
      && [[ -d "${curl_prefix::="$(brew --prefix 2> /dev/null)"/opt/curl}" ]]; then
  fpath=($curl_prefix/share/zsh/site-functions $fpath)
fi
unset curl_prefix
```

Proposed change (add a cache file):

```zsh
# Add completion for keg-only brewed curl when available.
if (( $+commands[brew] )); then
  # 'brew --prefix' is slow; cache its output
  cache_file="${XDG_CACHE_HOME:-$HOME/.cache}/prezto/brew-cache.zsh"
  if [[ "${ZDOTDIR:-$HOME}/.zpreztorc" -nt "$cache_file" || ! -s "$cache_file" ]]; then
    mkdir -p "$cache_file:h"
    echo "brew_prefix=$(brew --prefix 2> /dev/null)" >! "$cache_file"
  fi
  source "$cache_file"
  if [[ -d "${curl_site_functions::=$brew_prefix/opt/curl/share/zsh/site-functions}" ]]; then
    fpath=($curl_site_functions $fpath)
  fi
  unset cache_file brew_prefix curl_site_functions
fi
```

I'm open to discuss other solutions or implementations, but this implementation is similar to what the Perl module does so I modeled off that. I'm currently using it as a .patch file in my personal $ZDOTDIR and this sped up loads by more than 1/10th of a second.

### PR DESCRIPTION
**Pull request by indrajitr**
## Proposed Changes

For performance reasons, we prefer detecting Homebrew prefix internally instead of the more idiomatic form `brew --prefix`.
 
We attempt looking up $HOMEBREW_PREFIX or $HOMEBREW_REPOSITORY first (in case `brew shellenv` has been sourced-in earlier). Else, we look it up by resolving absolute path of $HOMEBREW_REPOSITORY. $HOMEBREW_PREFIX is
same as $HOMEBREW_REPOSITORY except when Homebrew is installed in '/usr/local' ($HOMEBREW_REPOSITORY == '/usr/local/Homebrew'). This is usually the case for Intel Macs.
 
This should work for most standard (and officially documented) Homebrew installations.
 
For implementation details in Homebrew, see: https://github.com/Homebrew/brew/blob/2a850e02d8f2dedcad7164c2f4b95d340a7200bb/bin/brew#L62-L70

Also, there are more details in Homebrew/brew#9177

Fixes #1978, #1979

### CODE REVIEW

**Comment by mattmc3** [comment]
This needs changed to `${commands[brew]:A:h:h}` to handle symlinked homebrew bin locations installs.

**Comment by indrajitr** [comment]
@mattmc3: Turns out, there is more to it :) Repo path works for M1 (or custom `brew` location), but not when it is in `/usr/local` on Intel Mac.

Can you please give this one a try?

**Comment by jeffwidman** [comment]
Thanks for working on this @indrajitr, nice improvement for sure!

File: modules/completion/init.zsh [hunk 1]
```diff
@@ -14,12 +14,16 @@ fi
...
 
-# Add completion for keg-only brewed curl when available.
-if (( $+commands[brew] )) \
-      && [[ -d "${curl_prefix::="$(brew --prefix 2> /dev/null)"/opt/curl}" ]]; then
-  fpath=($curl_prefix/share/zsh/site-functions $fpath)
+# Add completion for keg-only brewed curl on macOS when available.
+if (( $+commands[brew] )); then
+  brew_prefix=${HOMEBREW_PREFIX:-${HOMEBREW_REPOSITORY:-$commands[brew]:A:h:h}}
+  # $HOMEBREW_PREFIX defaults to $HOMEBREW_REPOSITORY but is explicitly set to
+  # /usr/local when $HOMEBREW_REPOSITORY is /usr/local/Homebrew.
```
Comment (jeffwidman) [inline modules/completion/init.zsh:21]: ? That's weird/confusing.

Comment (indrajitr) [reply modules/completion/init.zsh:21]: Which is why the implementation details in Homebrew is referred right there. See: https://github.com/Homebrew/brew/blob/2a850e02d8f2dedcad7164c2f4b95d340a7200bb/bin/brew#L62-L70

Comment (jeffwidman) [reply modules/completion/init.zsh:21]: Oh yeah, I totally thought you did the right thing, I just find it weird/confusing that the Homebrew project chose to do this.

**Review by jeffwidman — APPROVED** [review]

**Pull request merged**